### ISSUE TITLE
background methods with "void" type crash the app in 0.16.4

### ISSUE THREAD

**Issue by stopiccot** [calabash/calabash-ios#885]
I've upgraded calabash-ios gem to 0.16.4 a time ago, but ios server stayed 0.14.3. Despite gem-server-incompatible-versions warning everything was working fine. Today I've decided to upgrade ios framework  to 0.16.4. Now my app crashes at startup:

```
Process:               MemberFisikaliOS [28022]
Path:                  /Users/USER/Library/Developer/CoreSimulator/Devices/522BD82B-4FC4-4F64-A515-E6BDCF49EB95/data/Containers/Bundle/Application/55C40136-7258-454A-94C8-3CE76697CEEE/MemberFisikaliOS.app/MemberFisikaliOS
Identifier:            MemberFisikaliOS
Version:               3.1 (3.1.6)
Code Type:             X86 (Native)
Parent Process:        launchd_sim [27956]
Responsible:           MemberFisikaliOS [28022]
User ID:               501

Date/Time:             2015-10-25 20:12:55.689 +0300
OS Version:            Mac OS X 10.11.1 (15B42)
Report Version:        11
Anonymous UUID:        2D6B0298-5F6C-0CD3-298E-519D3C9AFFDC

Sleep/Wake UUID:       D88F11FE-D2B6-4D32-AE17-761F32F1B7B4

Time Awake Since Boot: 30000 seconds
Time Since Wake:       1200 seconds

System Integrity Protection: enabled

Crashed Thread:        0  Dispatch queue: com.apple.main-thread

Exception Type:        EXC_BREAKPOINT (SIGTRAP)
Exception Codes:       0x0000000000000002, 0x0000000000000000
Exception Note:        EXC_CORPSE_NOTIFY

Application Specific Information:
CoreSimulator 191.4 - Device: iPhone 5s - Runtime: iOS 9.1 (13B137) - DeviceType: iPhone 5s

Thread 0 Crashed:: Dispatch queue: com.apple.main-thread
0   com.apple.CoreFoundation      	0x08c678c4 __NSI0 + 612
1   com.apple.CoreFoundation      	0x08c91f5e __NSI1 + 318
2   com.apple.CoreFoundation      	0x08c91c1a -[NSInvocation getArgument:atIndex:] + 154
3   com.apple.CoreFoundation      	0x08d0d5f0 -[NSInvocation getReturnValue:] + 48
4   com.fisikal.enterprise.member.fisikal	0x000b69b1 -[LPBackdoorRoute JSONResponseForMethod:URI:data:] + 2065
5   com.fisikal.enterprise.member.fisikal	0x0009623a -[LPRouter httpResponseOnMainThreadForMethod:URI:] + 1962
6   com.fisikal.enterprise.member.fisikal	0x000967e9 __38-[LPRouter httpResponseForMethod:URI:]_block_invoke + 89
7   libdispatch.dylib             	0x091fd6fd _dispatch_client_callout + 14
8   libdispatch.dylib             	0x091e739b _dispatch_barrier_sync_f_slow_invoke + 66
9   libdispatch.dylib             	0x091fd6fd _dispatch_client_callout + 14
10  libdispatch.dylib             	0x091e6296 _dispatch_main_queue_callback_4CF + 689
```

**Comment by jmoody** [comment]
@stopiccot 

From the stack trace, I see you are trying to call a backdoor method.

To support Xcode 7 and bitcode we needed to update a bunch of libraries we depend on and this lead to changes in the backdoor route.

Can you paste the Objective-C and the ruby backdoor call into this case?

If you are using Objective-C message forwarding explicitly, then you will have to update your backdoor calls.

Specifically, if your UIApplicationDelegate is overriding:

```
respondsToSelector:
forwardingTargetForSelector:
```

to send messages to another object, you'll have to update your approach.

**Comment by jmoody** [comment]
> but ios server stayed 0.14.3

You must update the server manually.

https://github.com/calabash/calabash-ios/wiki/Updating-your-Calabash-iOS-version

**Comment by jmoody** [comment]
@Neil3079 

Please see above for why the changes to backdoor were necessary.

Can you paste your backdoor here?

Thanks for reporting!

**Comment by Neil3079** [comment]
sure thanks for the quick response I appreciate it! 

Ruby 
```
backdoor("networkStatus:",{offline:true})
```

Swift

```

func networkStatus(dict: NSDictionary) -> String {
        if dict["offline"] != nil {
            Injector.injector.getSimpleClass(NetworkMonitor).processOffline()
            return "App Offline"
        } else {
            Injector.injector.getSimpleClass(NetworkMonitor).processOnline()
            return "App Online"
        }
  }
```

This is one example as we have a lot of backdoors, all are crashing as far as I am able to tell

**Comment by Neil3079** [comment]
I don't seem to be able to recreate the issue in a new project only seems to happen if on upgraded projects as far as I can see

**Comment by jmoody** [comment]
My Swift is very weak:

```
Injector.injector.getSimpleClass(NetworkMonitor).processOffline()
```

That looks like a reflection call which is going cause problems.

If I am correct, you'll need to create an instance of NetworkMonitor and call the "processOffline()" and "processOnline()" methods of it.

I super sorry that the implementation broke but we had to upgrade a bunch of code to support bitcode.

**Comment by stopiccot** [comment]
Our app is Xamarin-based. We do not override `respondsToSelector` nor `forwardingTargetForSelector`

**Comment by jmoody** [comment]
@stopiccot What does your backdoor call look like?

**Comment by Neil3079** [comment]
No I'm afraid not this is our own implementation of an injection framework which doesn't work through reflection but rather protocols, backdoors without Injection code are just as likely to fail, even if i tear out everything and just return a string

**Comment by Neil3079** [comment]
![image](https://cloud.githubusercontent.com/assets/2815318/10759853/b19316d0-7cb1-11e5-9975-7bb8cb7605fc.png)

In case it helps this is where the exception is thrown

**Comment by jmoody** [comment]
@Neil3079 

Your error is a little different:

```
*** Terminating app due to uncaught exception 'NSInvalidArgumentException', 
reason: 
'+[NSInvocation _invocationWithMethodSignature:frame:]: 
method signature argument cannot be nil'
```

This means a valid method signature could not be created.  Hmm.

Can you whip up a sample Swift project that demonstrates this problem?

I have been testing the backdoor route against a Swift app, but I don't see this behavior.

**Comment by Neil3079** [comment]
I am currently working on that! unfortunately it seems to work with a fresh project

**Comment by stopiccot** [comment]
It's pretty obvious backdoor method

``` csharp
public abstract partial class AppDelegate : XFormsApplicationDelegate {

    [Action("setBaseURL:")]
    protected void SetBaseURL(NSString baseUrl) {
        FormsApplication.BackdoorSetBaseURL(baseUrl);
    }
}
```

**Comment by stopiccot** [comment]
Changing method implementation to

```csharp
protected void SetBaseURL(NSString baseUrl) {
    System.Diagnostics.Debug.WriteLine("SetBaseURL is called");
}
```

doesn't help

**Comment by stopiccot** [comment]
I'm not really familiar with bitcode yet. Maybe you have a calabash.framework with bitcode disabled? Tried to clone calabash-ios-server repo to 0.16.4 tag. But `make framework` fails for me. And for some reason there is no framework-type target in Xcode project which is really weird.

**Comment by stopiccot** [comment]
Missclick :)

**Comment by Neil3079** [comment]
Is there anyway we can examine the output of LPLogError in our project

**Comment by jmoody** [comment]
@stopiccot 

> make framework fails for me

How does it fail?  Please paste the error.

> there is no framework-type target in Xcode project which is really weird

We build releases from the command line, not from Xcode.

**Comment by jmoody** [comment]
@Neil3079 

> Is there anyway we can examine the output of LPLogError in our project

Yes.

```
# Tails the log file of the iOS Simulator that is currently open
$ bundle exec run-loop simctl tail
```

**Comment by stopiccot** [comment]
> We build releases from the command line, not from Xcode.
I still don't get the idea of a script building static-library-type Xcode target by running `xcodebuild` and then "assembling" it into framework by a series of mkdirs and lns. Why you can't simply `xcodebuild` framework directly?

> If you are a Xamarin, you must use the Test Cloud Agent? Correct?
Nope, Xamarin TestCloud is too limited for us so we use calabash "directly". We are bundling calabash.framework into our Xamarin app using Additional mtouch arguments option in Xamarin Studio.

**Comment by stopiccot** [comment]
And building error is:

```
INFO: Building ARM library for framework succeeded.
ditto: can't get real path for source 'build/framework/arm/Build/Intermediates/UninstalledProducts/libcalabash.a'
ERROR: Could not copy:
ERROR:   source: build/framework/arm/Build/Intermediates/UninstalledProducts/libcalabash.a
ERROR:   target: Products/framework/arm/libcalabash.a
ERROR: The source file does not exist
ERROR: Did a previous xcodebuild step fail?
ERROR: Exiting 1
make: *** [framework] Error 1
```

**Comment by jmoody** [comment]
> Why you can't simply xcodebuild framework directly?

When Calabash was first written, there was no way to build a framework for iOS.

> Nope, Xamarin TestCloud is too limited for us so we use calabash "directly". We are bundling calabash.framework into our Xamarin app using Additional mtouch arguments option in Xamarin Studio.

Gotcha.

There is (was?) a known issue with Xamarin Studio, Xcode 7, and bitcode.  I would be surprised, however, if this was the root cause of the backdoor crashes.

Based your output, however, I think you are running Xcode 6.  Is that correct?

> ditto: can't get real path for source

Weird.  We test this in two CI systems against Xcode 6 and Xcode 7.  I wonder what is going on.  Do you have tmate installed and would you mind doing a quick session?  http://tmate.io/

**Comment by Neil3079** [comment]
No output from ```bundle exec run-loop simctl tail```, really at a loss with this one

**Comment by jmoody** [comment]
@Neil3079

> No output from bundle exec run-loop simctl tail, really at a loss with this one

No output at all or there is nothing in the logs?

The simulator you are targeting must be open.

If you can't find the log with run-loop:

```
$ export SIM_UUID=5FF7F09E-2371-4E88-B562-3A098D96B2C7
$ tail -f ~/Library/Logs/CoreSimulator/${SIM_UUID}/system.log
```

**Comment by Neil3079** [comment]
There is stuff in the logs but nothing when I call the backdoor method crashing the app

**Comment by jmoody** [comment]
@Neil3079 If you are having trouble reproducing you can send me .app for the simulator _that is not linked with calabash.framework_ and I can take a look using dylib injection.

**Comment by stopiccot** [comment]
> Based your output, however, I think you are running Xcode 6. Is that correct?

Nope, it's Xcode 7 on 10.11

**Comment by stopiccot** [comment]
Changing
`ARM_LIBRARY_XC7="${ARM_BUILD_DIR}/Build/Intermediates/ArchiveIntermediates/${XC_TARGET}/IntermediateBuildFilesPath/UninstalledProducts/${LIBRARY_NAME}"`
to
`ARM_LIBRARY_XC7="${ARM_BUILD_DIR}/Build/Intermediates/ArchiveIntermediates/${XC_TARGET}/IntermediateBuildFilesPath/UninstalledProducts/iphoneos/${LIBRARY_NAME}"`
fixed build for me

**Comment by jmoody** [comment]
That is very strange.  We target that Xcode in CI and it is what I have installed locally.

**Comment by stopiccot** [comment]
Btw, can I use NSLogs in calabash.framework?

**Comment by jmoody** [comment]
> NSLogs

We are migrating away from NSLog because it is known to clobber UIAutomation.

We are using a namespaced CocoaLumberjack - LPCocoaLumberjack - for logging.

**Comment by stopiccot** [comment]
Replicated in a sample project https://github.com/stopiccot/CalabashBugXamarinFormsSampleProject
Running `backdoor "TestBackdoor:", nil` from `calabash-ios console` crashes the app.
The most interesting thing is that backdoor method is actually called. It looks like app crashes when returning from backdoor method. 

Anyway hope this sample project helps you to figure things out cause I have no idea what is going on)

**Comment by Neil3079** [comment]
Hi all sorry I had to duck out last night, reading through and catching up it seems there could be multiple issues as in my case it definitely doesn't execute the code inside the backdoor method and actually crashes when creating the NSInvocation @jmoody is there a way to attach the calabash server to my project as a submodule or something so I can step through the code?

**Comment by jmoody** [comment]
@stopiccot 

We have reproduced.  We think it might be an issue with Forms apps.  We are investigating.  I might escalate this to Xamarin support.  I will keep you posted.

@Neil3079 

I agree there are two problems.

> is there a way to attach the calabash server to my project as a submodule or something so I can step through the code?

There are a few ways:

1. Use lldb on the command line.  You are on your own.  I do this occasionally, but I am not an lldb expert so I have no advice.
2. Add the server project sources to your project.   It is not pretty so I tend to avoid it, but other devs on the team do this all the time.
3. Use dylib injection and logging - this what I usually do.  Unfortunately, dylib injection is broken in run-loop - the fix is in the develop branch of run-loop.

**Comment by Neil3079** [comment]
@jmoody I have finaly solved the issue I was having and it was not a calabash issue, a third party framework was doing something strange that resulted in when you call [[UIApplication sharedApplication] delegate] (as you do in the calabash framework) it was returning not my AppDelegate but a proxy object that contains a reference to my app delegate and its own delegate

TLDR;

Not calabash issue, third party framework doing some weird stuff, solved by not initialising that framework when running calabash tests.

Thanks for all the help with this

**Comment by jmoody** [comment]
@Neil3079 Whew.

Would you mind sharing with me the framework?  You can contact me privately if you want.

**Comment by stopiccot** [comment]
Any ideas what Xamarin Forms are doing wrong? Maybe it could be somehow workarounded for now?

**Comment by jmoody** [comment]
@stopiccot 

> Any ideas what Xamarin Forms are doing wrong?

It is not a question of anyone doing anything wrong.  And it was just a guess because I have seen a problem with NSInvocation in Calabash + Forms in the past.

Your example fails because the return type of the method is void.

0.14.3 supported void type returns, but 0.16.4 does not.  This is an oversight (bug) on my part.  Apologies.

The rules for backdoors in 0.16.4 are summarized here:

https://github.com/calabash/ios-smoke-test-app/blob/master/CalSmokeApp/features/backdoor.feature

The good news is that 0.17.0 will support arbitrary signatures.  I am working through the backlog for a 0.17.0 release.

**Expand backdoor route to handle arbitrary method signatures** #220

Since you build the server from sources and appear to be comfortable in ruby, you could build the server from the `develop` branch and add the patch mentioned in #220 to the gem - be sure to patch the `develop` branch of the gem.  That would give you arbitrary signatures in advance of 0.17.0.

Otherwise, update your backdoors to return NSString (or some other supported type).

**Comment by stopiccot** [comment]
I've updated backdoor methods to return NSString and everything works fine now. Thanks for help.

**Comment by jmoody** [comment]
@stopiccot Sorry about the confusion and the run around.  I should have been able to identify the problem instantly.  Thank you for reproducing.
